android-components/components/service/contile/src/main/java/mozilla/components/service/contile/ContileTopSitesProvider.kt

```diff
@@ -98,12 +105,21 @@ class ContileTopSitesProvider(
             if (response.isSuccess) {
                 val responseBody = response.body.string(Charsets.UTF_8)
 
+                if (response.status == Response.NO_CONTENT) {
+                    // If the response is 204, we should invalidate the cached top sites
+                    cacheState = cacheState.invalidate()
+                    getCacheFile().delete()
+                    return listOf()
+                }
+
                 return try {
-                    JSONObject(responseBody).getTopSites().also {
-                        if (maxCacheAgeInMinutes > 0) {
-                            writeToDiskCache(responseBody)
-                        }
-                    }
+                    val jsonBody = JSONObject(responseBody)
+                    writeToDiskCache(
+                        response.headers.computeValidFor() * DateUtils.SECOND_IN_MILLIS,
+                        jsonBody.getJSONArray("tiles"),
```

Comment:
nit: I think `"tiles"` could be `CACHE_TOP_SITES_KEY` here as well. 

It also feels a little funky to me that we are calling `jsonBody.getJSONArray("tiles")` here and then immediately doing basically the same thing with `jsonBody.getTopSites()`. I can't immediately think of a clean way to resolve that so I think we can keep it as is unless something comes to your mind.